### PR TITLE
Add openshift.io/deployer-pod.type label

### LINKED ISSUE #8498: Provide a way to distinguish hook pods from other deployer pods
All deployer pods (the strategy pod and hook pods) are assigned an `openshift.io/deployer-pod-for.name` label so they can be identified as deployers. However, it's also useful to further identify these pods by their specific role. For example, a label like `openshift.io/deployer-pod.type=strategy` or `openshift.io/deployer-pod.type=pre-hook`.

### PR DESCRIPTION
**Pull request by legionus**
Fix #8498 

@kargakis please review

### CODE REVIEW

**Comment by legionus** [comment]
[test]

**Comment by kargakis** [comment]
Looks good for the description of the issue but I didn't realize at first we wanted to filter even between different types of hooks cc: @smarterclayton @ironcladlou

File: test/fixtures/hooks-dc.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+apiVersion: v1
+kind: DeploymentConfig
+metadata:
+  name: failing-dc-hooks
```
Comment (kargakis) [inline test/fixtures/hooks-dc.yaml]: Rename this to "complete-dc-hooks" since they do complete

File: test/end-to-end/core.sh [hunk 2]
```diff
@@ -184,6 +184,11 @@ os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'Usin
 os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'kubectl-with-token'
 
 echo "[INFO] Testing deployment logs and failing pre and mid hooks ..."
+# test hook selectors
+oc create -f test/fixtures/hooks-dc.yaml
```
Comment (kargakis) [inline test/end-to-end/core.sh]: Use `os::cmd::expect_success` otherwise @stevekuznetsov will haunt you:)

Comment (kargakis) [inline test/end-to-end/core.sh]: Also since we don't use this deployment anymore, delete it after you are done testing.

**Comment by legionus** [comment]
re[test]

**Comment by kargakis** [comment]
LGTM

File: test/end-to-end/core.sh [hunk 3]
```diff
@@ -184,6 +184,11 @@ os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'Usin
 os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'kubectl-with-token'
 
 echo "[INFO] Testing deployment logs and failing pre and mid hooks ..."
+# test hook selectors
+os::cmd::expect_success 'oc create -f test/fixtures/complete-dc-hooks.yaml'
+wait_for_command 'oc get pods -l openshift.io/deployer-pod.type=hook-pre  | grep -q Completed' $((60*TIME_SEC))
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: These tests should be written as follows:

```sh
os::cmd::try_until_text 'oc get pods -l openshift.io/deployer-pod.type=hook-pre --jsonpath=<path to field>' 'completed'
```

This way we:
1. are consistent in using `os:;cmd`
2. get jUnit integration
3. don't depend on unstable `oc get

File: test/end-to-end/core.sh [hunk 4]
```diff
@@ -184,6 +184,11 @@ os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'Usin
 os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'kubectl-with-token'
 
 echo "[INFO] Testing deployment logs and failing pre and mid hooks ..."
+# test hook selectors
+os::cmd::expect_success 'oc create -f test/fixtures/complete-dc-hooks.yaml'
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: The path to the fixture file should be written relative to `${OS_ROOT}` for stability.

**Comment by legionus** [comment]
@kargakis @stevekuznetsov Fixed.

File: test/end-to-end/core.sh [hunk 5]
```diff
@@ -184,6 +184,11 @@ os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'Usin
 os::cmd::expect_success_and_text "cat '${LOG_DIR}/kubectl-with-token.log'" 'kubectl-with-token'
 
 echo "[INFO] Testing deployment logs and failing pre and mid hooks ..."
+# test hook selectors
+os::cmd::expect_success "oc create -f ${OS_ROOT}/test/fixtures/complete-dc-hooks.yaml"
+os::cmd::try_until_text 'oc get pods -l openshift.io/deployer-pod.type=hook-pre  -o jsonpath={.items[*].status.phase}' 'Succeeded'
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: With `.items[*].status.phase` you will get a list of phases, and your assert is that at least one of these phases is `Succeeded` -- is that what we want? At least one pod has succeeded? Do we want to know that all of them have?

Comment (legionus) [inline test/end-to-end/core.sh]: I thought that the second argument is the whole output.
Should I use `^Succeeded$` or something for that ?

Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: It's given to `grep` so you would need line anchors if you wanted them. I believe the `*` notation will give you `[Complete]` if there's only one pod.

Comment (legionus) [inline test/end-to-end/core.sh]: No. I'll get a list of values in line. I want to be sure that I found only one pod. That's why I use `^Succeeded$`.

**Comment by legionus** [comment]
re[test]

File: pkg/deploy/strategy/support/lifecycle_test.go [hunk 6]
```diff
@@ -274,6 +275,7 @@ func TestHookExecutor_makeHookPod(t *testing.T) {
 				ObjectMeta: kapi.ObjectMeta{
 					Name: namer.GetPodName(deploymentName, "hook"),
 					Labels: map[string]string{
+						deployapi.DeploymentPodTypeLabel:        "hook",
```
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle_test.go]: Why these labels?

Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle_test.go]: oh, I see. Nevermind

**Comment by kargakis** [comment]
[merge]

**Pull request merged**